### PR TITLE
[halide-backend] Initial implementation of HalideKernel and HalideScheduling

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129321
* #129320
* #129036
* #127506
* #129026
* #129025
* __->__ #126417

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by mitkotak** [comment]
Hey thanks for the work ! Was wondering if there's plans to support GPUs as well ? Thanks !

**Comment by jansel** [comment]
> Hey thanks for the work ! Was wondering if there's plans to support GPUs as well ? Thanks !

Yes, I plan to add GPU support after finishing work on initial functionality/correctness (this PR).

> I'll bump ￼@mitkotak 's question for ￼@jansel and ask for something slightly different. In particular, I think it would be nice if it was easy for me to plug Halide Generators into Torch via this infrastructure and have them be GPU capable if I write a GPU schedule. I think GPU is probably not a good idea if this has to be fully automatic given the state of the Halide GPU auto-schedulers, but I think being able to write our own GPU kernels in Halide and have them be easily useable in torch via the infrastructure here that would be great. Is anything like this a design goal? If it isn't, could we convince/help out?

@wraith1995 user-define Halide kernels is not the goal of the PR stack.  The focus here is generating Halide kernels as a backend for TorchInductor.

That said, you could likely implement user-defined Halide kernels somewhat easily based on #126416 -- which involves AOT compilation and fast torch/python bindings.  The PR still needs GPU support (mainly just plumbing through the user-context), but it is already relatively close.

File: torch/_inductor/codegen/halide.py [hunk 1]
```diff
@@ -0,0 +1,1121 @@
...
+            def generate(g):
+        """
+        )
+        code.do_indent()
+        for _, arg in self.halide_argdefs():
+            code.writeline(f"{arg.name} = g.{arg.name}")
+        for old, new in self.args.aliases():
+            code.writeline(f"{old} = {new}")
+
+        dom_size = {}
```
Comment (shunting314) [inline torch/_inductor/codegen/halide.py:988]: Still need some time to review. But what the difference between the concept of 'Dom' here to a 'dimension' of a buffer?

Comment (jansel) [reply torch/_inductor/codegen/halide.py:988]: Halide has a concept of a RDom: https://halide-lang.org/docs/class_halide_1_1_r_dom.html

You use hl.RDom for reductions or updates and hl.Var for pure definitions.  RDom is more flexible, so I use that in some cases.

File: torch/_inductor/codegen/halide.py [hunk 2]
```diff
@@ -0,0 +1,1121 @@
...
+            return index_str, value_str
+
+        self.needs_dom_indexing = True
+        # Fall back to using RDom-style store
+        self.body.writeline(
+            DeferredLine(name, f"{var}[hl.Var()] = hl.undef({var}.type())")
+        )
+        suffix = "idom" if self.inside_reduction else "odom"
+        value_str = value.with_dom(suffix)
+        index_str = self.index_to_str(self.index_to_dom(index, suffix))
```
Comment (shunting314) [inline torch/_inductor/codegen/halide.py:741]: Why different hl.Dom is used for the code inside a reduction and the code outside of a reduction?

**Review by shunting314 — APPROVED** [review]

**Comment by shunting314** [comment]
Can we have a paste of the generated wrapper to show how the result of halide codegen looks like?

File: torch/_inductor/codegen/halide.py [hunk 3]
```diff
@@ -0,0 +1,1121 @@
...
+            code.writeline(line)
+        code.writeline("")
+        code.writeline("assert g.using_autoscheduler()")
+
+        for _, arg in self.halide_argdefs():
+            # fallback=1 below because halide requires buffers to be at least as large as the estimates
+            # This causes crashes if our estimate is greater than the vector length
+            # https://github.com/halide/Halide/issues/3103
+            if isinstance(arg, SizeArg):
+                hint = V.graph.sizevars.size_hint(arg.expr, fallback=1)
```
Comment (eellison) [inline torch/_inductor/codegen/halide.py:1034]: could use var range lower bound instead

**Review by eellison — APPROVED** [review]
agreed, would be helpful to have some reference generated halide kernels

File: test/inductor/test_torchinductor.py [hunk 4]
```diff
@@ -1131,6 +1154,7 @@ def fn(a):
         self.common(fn, [torch.linspace(-10, 10, 41)])
 
     @skipCUDAIf(not SM80OrLater, "uses bfloat16 which requires SM >= 80")
+    @skip_if_halide  # bf16
```
[the remark this answers is not in the document]
Comment (jansel) [reply test/inductor/test_torchinductor.py:1157]: It just isn't exposed in the Python API.  I actually already submitted a fix to add it.

File: torch/_inductor/codegen/halide.py [hunk 5]
```diff
@@ -0,0 +1,1121 @@
...
+
+class Unsupported(RuntimeError):
+    def __init__(self, thing):
+        super().__init__(f"halide backend does not support: {thing}")
+
+
+class HalidePrinter(PythonPrinter):
+    @staticmethod
+    def cast_index(expr):
+        return f"hl.cast({V.kernel.index_dtype}, {expr})"
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_inductor/codegen/halide.py:69]: yeah...

**Comment by jansel** [comment]
Here is some example output:
```py
# Source Nodes: [add, sum_1], Original ATen: [aten.add, aten.sum]
# add => add
# sum_1 => sum_1
halide_kernel_0 = async_compile.halide(HalideMeta(argtypes=[HalideInputSpec(ctype='float*', name='in_ptr0', shape=['64L']), HalideInputSpec(ctype='float*', name='in_ptr1', shape=['64L']), HalideInputSpec(ctype='float*', name='out_ptr0', shape=['8L'])], target='host-strict_float-no_runtime-no_asserts', scheduler='Adams2019', scheduler_flags={'parallelism': 8}, cuda_device=None), '''
import halide as hl

@hl.generator(name="kernel")
class Kernel:
    in_ptr0 = hl.InputBuffer(hl.Float(32), 1)
    in_ptr1 = hl.InputBuffer(hl.Float(32), 1)
    out_ptr0 = hl.OutputBuffer(hl.Float(32), 1)

    def generate(g):
        in_ptr0 = g.in_ptr0
        in_ptr1 = g.in_ptr1
        out_ptr0 = g.out_ptr0
        xindex = hl.Var('xindex')
        rindex = hl.Var('rindex')
        r1 = rindex
        x0 = xindex
        idom = hl.RDom([hl.Range(0, 8), hl.Range(0, 8)])
        odom = hl.RDom([hl.Range(0, 8)])
        rdom = hl.RDom([hl.Range(0, 8)])
        xindex_idom = idom.x
        xindex_odom = odom.x
        rindex_idom = idom.y
        r1_idom = rindex_idom
        x0_idom = xindex_idom
        x0_odom = xindex_odom
        tmp0 = hl.Func('tmp0')
        tmp0[rindex, xindex] = in_ptr0[r1 + (8*x0)]
        tmp1 = hl.Func('tmp1')
        tmp1[rindex, xindex] = in_ptr1[r1 + (8*x0)]
        tmp2 = hl.Func('tmp2')
        tmp2[rindex, xindex] = tmp0[rindex, xindex] + tmp1[rindex, xindex]
        tmp3 = hl.Func('tmp3')
        tmp3[xindex] = hl.sum(rdom, tmp2[rdom, xindex])
        out_ptr0[x0] = hl.cast(out_ptr0.type(), tmp3[xindex])

        assert g.using_autoscheduler()
        in_ptr0.set_estimates([hl.Range(0, 64)])
        in_ptr1.set_estimates([hl.Range(0, 64)])
        out_ptr0.set_estimates([hl.Range(0, 8)])

if __name__ == "__main__":
    hl.main()
else:
    hl.load_plugin('/home/jansel/conda/envs/pytorch/lib/libautoschedule_adams2019.so')
    target = hl.Target('host-strict_float-no_runtime-no_asserts')
    autoscheduler = hl.AutoschedulerParams('Adams2019', {'parallelism': 8})
    with hl.GeneratorContext(target, autoscheduler):
        gen = Kernel()
        pipeline = gen._build_pipeline()
        # gen.compile_to_callable() does not run the autoscheduler
        pipeline.apply_autoscheduler(target, autoscheduler)
        kernel = pipeline.compile_to_callable([
                gen._get_input_parameter(a.name)._to_argument()
                for a in gen._get_arginfos()
                if a.dir == hl.ArgInfoDirection.Input
            ], target)
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, arg1_1 = args
    args.clear()
    assert_size_stride(arg0_1, (8, 8), (8, 1))
    assert_size_stride(arg1_1, (8, 8), (8, 1))
    buf0 = empty_strided_cpu((8, ), (1, ), torch.float32)
    halide_kernel_0(arg0_1, arg1_1, buf0)
    run_intermediate_hooks('sum_1', buf0)
    del arg0_1
    del arg1_1
    return (buf0, )

```

File: torch/_inductor/codegen/halide.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/codegen/halide.py:753]: They are different sizes.  One is rnumel, one is xnumel, another is rnumel*xnumel.

**Pull request closed without merging**